java/arcs/core/data/expression/Evaluator.kt

```diff
@@ -67,27 +67,11 @@ class ExpressionEvaluator(
     override fun <T> visit(expr: Expression.ObjectLiteralExpression<T>): Any = expr.value as Any
 
     override fun <E, T> visit(expr: Expression.FromExpression<E, T>): Any {
-        var sequence = expr.expr.accept(this)
-
-        if (sequence is List<*>) {
-            sequence = sequence.asSequence()
-        }
-
-        require(sequence is Sequence<*>) {
-            "From source expression of type ${sequence::class} cannot be converted to a Sequence"
-        }
-
-        val fromSequence = sequence.map { value ->
-            currentScope.set(expr.iterationVar, value as Any)
-            value
-        }
-
-        return if (expr.qualifier != null) {
-            (expr.qualifier.accept(this) as Sequence<T>).flatMap {
-                fromSequence
+        return (expr.qualifier?.accept(this) as Sequence<T>? ?: sequenceOf(null)).flatMap {
+            asSequence<T>(expr.expr.accept(this)).map { value ->
```

Comment:
Nice trick, I though of doing something like this but for some reason convinced myself the if/else was better, but that forced me to pre-declare the fromSequence to avoid duplication. 👍